### PR TITLE
Faction plugin

### LINKED ISSUE #770: Get/Set Faction Id - faction management
In previous releases of nwnx (nwnx_cool or funcs I think) - I think there was Get/Set faction id - for npcs/objects.
I know that in EE there is a ChangeFaction method - which accepts two npc's as the arguments - but this does not help if you want to manage factions outside of the NWScript code base.
Eg: Create your own faction management system in say .Net Core etc.

Also - use of ChangeFaction forces people down the route of using the age old 'faction areas' where you have to keep a member of each faction alive in isolation (so they don't kill each other) - just so you have an object reference in order to interact with their faction.

`object oFirstFac = GetObjectByTag("vampire_faction");
object oTarget = ...;
ChangeFaction(oTarget,oFirstFac);
`

Where as you could just as easily :

`
NWNX_Factions_SetFaction(oTarget,nVampireFactionId);
`

Eg: You could store off the int for the vampire faction for use later, instead of the idea of having to constantly get the object for the npc's faction.

Im not sure if this would be as simple as exposing the m_nFactionId on the faction object itself, or if it would be safer to go through the internal methods etc

void AddMember(OBJECT_ID oidNewMemberID, BOOL bInPostProcess = false);
BOOL AttemptRemove(OBJECT_ID oidRemoveeID, OBJECT_ID oidRemoverID);

Eg: 
Remove from current faction,
Add to new faction.

Or would the Add method take care of the remove?

Note sure if this would be ideal as a new plugin or could just go into the Objects plugin - since factions exist at object level and above. (At least for placeables,npcs, triggers,encounters - not sure about sound,waypoints,items etc)

### PR DESCRIPTION
**Pull request by Baaleos**
resolves #770 

Allows for getting and setting of a faction id without having to use a reference creature.
Means PW owners can output the faction ids to integer variables or databases etc

Tested in the following scenario:

```
result = true;
                Area ar = new Area(area);
                uint cre = NWScript.CreateObject(NWScript.OBJECT_TYPE_CREATURE, resref, ar.RandomLocationInArea(), 0, "");
                uint wolf = NWScript.CreateObject(NWScript.OBJECT_TYPE_CREATURE,"nw_wolf",ar.RandomLocationInArea(),0,"");
                int iCommoner = -1;
                int iWolf = -1;
                iCommoner = NWNX.Factions.GetFaction(cre);
                iWolf = NWNX.Factions.GetFaction(wolf);
                if (iCommoner == iWolf)
                {
                    result = false;
                    this.Note += "Commoner and Wolf were on same factions ";
                }
                NWNX.Factions.SetFaction(wolf,iCommoner);

                iWolf = NWNX.Factions.GetFaction(wolf);
                if (iCommoner != iWolf)
                {
                    result = false;
                    this.Note += "Commoner and Wolf were not on same faction ";
                }
                NWScript.DestroyObject(cre, 0.25f);
                NWScript.DestroyObject(wolf, 0.25f);
```

Greenshot 2020-02-28 15 17 46

If anyone has any suggestions on cleaning up the code etc - let me know.

### CODE REVIEW

**Comment by Baaleos** [comment]
Greenshot 2020-02-28 16 29 37


Updated branch - Faction functionality now moved to Creature plugin and is functioning as expected.

**Comment by Baaleos** [comment]
Updated to use 
```
if (auto *pCreature = creature(args))
    {
```

File: Plugins/Creature/Creature.cpp [hunk 1]
```diff
@@ -1876,4 +1882,38 @@ ArgumentStack Creature::SetDisarmable(ArgumentStack&& args)
...
+{
+    int32_t retVal = -1;
+    if (auto *pCreature = creature(args)){
+        const auto nFacId = Services::Events::ExtractArgument<int32_t>(args);
+        auto* pFaction= Globals::AppManager()->m_pServerExoApp->m_pcExoAppInternal->m_pFactionManager->GetFaction(nFacId);
+        if (pFaction){
+            pFaction->AddMember(pCreature->m_idSelf);
+            retVal = 1;
+        }else{
+            LOG_NOTICE("NWNX_Faction function called with invalid faction id");
```
Comment (mtijanic) [inline Plugins/Creature/Creature.cpp]: This is no longer a `NWNX_Faction` function. But you probably don't need the print anyway.
Same for other two prints.

File: Plugins/Creature/Creature.cpp [hunk 2]
```diff
@@ -1876,4 +1882,38 @@ ArgumentStack Creature::SetDisarmable(ArgumentStack&& args)
     }
     return Services::Events::Arguments();
 }
+
+ArgumentStack Creature::SetFaction(ArgumentStack&& args)
+{
+    int32_t retVal = -1;
+    if (auto *pCreature = creature(args)){
```
Comment (mtijanic) [inline Plugins/Creature/Creature.cpp]: Style nit: `{` and `}` on own lines. Also, generally avoiding prefixes (except `p-`) on variables in nwnx, so `factionId` or something.

File: Plugins/Creature/NWScript/nwnx_creature.nss [hunk 3]
```diff
@@ -1603,4 +1614,21 @@ void NWNX_Creature_SetSpecialization(object creature, int class, int school)
     NWNX_PushArgumentObject(NWNX_Creature, sFunc, creature);
 
     NWNX_CallFunction(NWNX_Creature, sFunc);
+}
+
+int NWNX_Creature_GetFaction(object oCreature)
+{
+    string sFunc = "GetFaction";
+    NWNX_PushArgumentObject(NWNX_Creature, sFunc, oItem);
```
Comment (mtijanic) [inline Plugins/Creature/NWScript/nwnx_creature.nss]: What's `oItem`? Copy/paste issues? Same below.

File: Plugins/Creature/Creature.cpp (hunk 2 above)
Comment (Baaleos) [reply Plugins/Creature/Creature.cpp]: Done

File: Plugins/Creature/Creature.cpp (hunk 1 above)
Comment (Baaleos) [reply Plugins/Creature/Creature.cpp]: Corrected log message

**Comment by Baaleos** [comment]
Didn't realize there was a conflict on the change-log - should be resolved now

File: Plugins/Creature/Creature.cpp [hunk 4]
```diff
@@ -1876,4 +1882,46 @@ ArgumentStack Creature::SetDisarmable(ArgumentStack&& args)
...
+    return Services::Events::Arguments(retVal);
+}
+
+
+ArgumentStack Creature::GetFaction(ArgumentStack&& args)
+{
+    int32_t retVal = -1;
+    if(auto *pCreature = creature(args))
+    {
+        auto *faction = pCreature->GetFaction();
```
Comment (Daztek) [inline Plugins/Creature/Creature.cpp]: I'd change this to `if (auto *pFaction = pCreature->GetFaction())` on the off chance GetFaction() returns a nullptr.

File: Plugins/Creature/Creature.cpp [hunk 5]
```diff
@@ -1876,4 +1882,46 @@ ArgumentStack Creature::SetDisarmable(ArgumentStack&& args)
...
+        }
+        else
+        {
+            LOG_NOTICE("NWNX_Creature function called with invalid faction id");
+        }
+    }
+    else
+    {
+        LOG_NOTICE("NWNX_Creature function called on invalid creature");
+    }
```
Comment (Daztek) [inline Plugins/Creature/Creature.cpp]: This one isn't really needed, imo, none of the other functions have it either. 

Same for the one in GetFaction()

File: Plugins/Creature/Creature.cpp [hunk 6]
```diff
@@ -1876,4 +1882,46 @@ ArgumentStack Creature::SetDisarmable(ArgumentStack&& args)
...
+        const auto factionid = Services::Events::ExtractArgument<int32_t>(args);
+        auto* pFaction= Globals::AppManager()->m_pServerExoApp->m_pcExoAppInternal->m_pFactionManager->GetFaction(factionid);
+        if (pFaction)
+        {
+            pFaction->AddMember(pCreature->m_idSelf);
+            retVal = 1;
+        }
+        else
+        {
+            LOG_NOTICE("NWNX_Creature function called with invalid faction id");
```
Comment (Daztek) [inline Plugins/Creature/Creature.cpp]: I'd change this to `NWNX_Creature_SetFaction called with invalid faction id`

**Comment by mtijanic** [comment]
I'm gonna merge this now, and additional fixes can come in a new PR if needed.

**Comment by Baaleos** [comment]
thanks

**Pull request merged**